### PR TITLE
Fix dispatcher injecting

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️

### CODE REVIEW

File: src/Group.php [hunk 1]
```diff
@@ -59,18 +59,15 @@ public static function create(?string $prefix = null, $routes = [], DispatcherIn
         return new self($prefix, $callback, $dispatcher);
     }
 
-    public function withDispatcher(DispatcherInterface $dispatcher): self
+    public function injectDispatcher(DispatcherInterface $dispatcher): void
```
Comment (samdark) [inline src/Group.php:62]: Can it be immutable?

File: src/Route.php [hunk 2]
```diff
@@ -23,24 +23,26 @@ final class Route
...
+    private array $middlewareDefinitions = [];
     private array $defaults = [];
 
     private function __construct(?DispatcherInterface $dispatcher = null)
     {
         $this->dispatcher = $dispatcher;
     }
 
-    public function withDispatcher(DispatcherInterface $dispatcher): self
+    public function injectDispatcher(DispatcherInterface $dispatcher): void
```
Comment (samdark) [inline src/Route.php:34]: Can it be immutable?

Comment (yiiliveext) [reply src/Route.php:34]: > Can it be immutable?

No, it doesn't. We should reuse the stack on the next request in the request loop.

File: src/Group.php [hunk 3]
```diff
@@ -59,11 +59,11 @@ public static function create(?string $prefix = null, $routes = [], DispatcherIn
         return new self($prefix, $callback, $dispatcher);
     }
 
-    public function withDispatcher(DispatcherInterface $dispatcher): self
+    public function withDispatcher(MiddlewareDispatcher $dispatcher): self
     {
         $group = clone $this;
         $group->dispatcher = $dispatcher;
-        foreach ($group->items as $index => $item) {
+        foreach ($this->items as $index => $item) {
```
Comment (samdark) [inline src/Group.php:66]: The change looks suspicious.

**Comment by samdark** [comment]
👍

**Pull request merged**